[BUG] Saving user map preferences

When preferences are changed on the map 
- Save Route Segments
- Show Neighbor Info
- etc ...

The chosen preferences are not saved:

Docker container log:

[ERROR] Error saving user map preferences: SqliteError: table user_map_preferences has no column named created_at
    at Database.prepare (/app/node_modules/better-sqlite3/lib/methods/wrappers.js:5:21)
    at UserModel.saveMapPreferences (file:///app/dist/server/models/User.js:410:34)
    at DatabaseService.saveMapPreferencesAsync (file:///app/dist/services/database.js:8264:28)
    at file:///app/dist/server/server.js:4733:31
    at Layer.handleRequest (/app/node_modules/router/lib/layer.js:152:17)
    at next (/app/node_modules/router/lib/route.js:157:13)
    at file:///app/dist/server/auth/authMiddleware.js:76:13
    at process.processTicksAndRejections (node:internal/process/task_queues:104:5) {
  code: 'SQLITE_ERROR'



* Host OS: [Linux]
* Deployment type: Docker
  * Container version: 29.0.1

--- a/src/db/repositories/misc.ts
+++ b/src/db/repositories/misc.ts
@@ -1160,6 +1160,123 @@ export class MiscRepository extends BaseRepository {
       return [];
     }
   }
+
+  // ============ USER MAP PREFERENCES ============
+
+  /**
+   * Get map preferences for a user
+   */
+  async getMapPreferences(userId: number): Promise<Record<string, any> | null> {
+    try {
+      const userIdCol = this.dbType === 'postgres' ? '"userId"' : 'userId';
+      const rows = await this.executeQuery(
+        sql.raw(`SELECT * FROM user_map_preferences WHERE ${userIdCol} = ${userId} LIMIT 1`)
+      );
+
+      if (rows.length === 0) return null;
+      const row = rows[0] as any;
+
+      return {
+        mapTileset: row.map_tileset ?? row.mapTileset ?? null,
+        showPaths: Boolean(row.show_paths ?? row.showPaths ?? false),
+        showNeighborInfo: Boolean(row.show_neighbor_info ?? row.showNeighborInfo ?? false),
+        showRoute: Boolean(row.show_route ?? row.showRoute ?? true),
+        showMotion: Boolean(row.show_motion ?? row.showMotion ?? true),
+        showMqttNodes: Boolean(row.show_mqtt_nodes ?? row.showMqttNodes ?? true),
+        showMeshCoreNodes: Boolean(row.show_meshcore_nodes ?? row.showMeshCoreNodes ?? true),
+        showAnimations: Boolean(row.show_animations ?? row.showAnimations ?? false),
+        showAccuracyRegions: Boolean(row.show_accuracy_regions ?? row.showAccuracyRegions ?? false),
+        showEstimatedPositions: Boolean(row.show_estimated_positions ?? row.showEstimatedPositions ?? false),
+        positionHistoryHours: row.position_history_hours ?? row.positionHistoryHours ?? null,
+      };
+    } catch (error) {
+      logger.error('[MiscRepository] Failed to get map preferences:', error);
+      return null;
+    }
+  }
+
+  /**
+   * Save map preferences for a user (upsert).
+   * Uses raw SQL via Drizzle's sql template to avoid referencing columns that
+   * may not exist across all upgrade paths (the schema varies by migration state).
+   */
+  async saveMapPreferences(userId: number, preferences: {
+    mapTileset?: string;
+    showPaths?: boolean;
+    showNeighborInfo?: boolean;
+    showRoute?: boolean;
+    showMotion?: boolean;
+    showMqttNodes?: boolean;
+    showMeshCoreNodes?: boolean;
+    showAnimations?: boolean;
+    showAccuracyRegions?: boolean;
+    showEstimatedPositions?: boolean;
+    positionHistoryHours?: number | null;
+  }): Promise<void> {
+    try {
+      // Use the userId column name based on database type
+      // SQLite baseline uses userId (camelCase), PG/MySQL also use userId
+      const userIdCol = this.dbType === 'postgres' ? '"userId"' : 'userId';
+
+      // Check if preferences exist
+      const existing = await this.executeQuery(
+        sql.raw(`SELECT id FROM user_map_preferences WHERE ${userIdCol} = ${userId} LIMIT 1`)
+      );
+      const hasExisting = existing.length > 0;
+
+      if (hasExisting) {
+        // Build dynamic UPDATE — only set columns that were provided
+        // PG uses boolean type (true/false), SQLite/MySQL use integer (1/0)
+        const bv = (v: boolean) => this.dbType === 'postgres' ? String(v) : (v ? '1' : '0');
+        const sets: string[] = [];
+        if (preferences.mapTileset !== undefined) sets.push(`map_tileset = '${preferences.mapTileset.replace(/'/g, "''")}'`);
+        if (preferences.showPaths !== undefined) sets.push(`show_paths = ${bv(preferences.showPaths)}`);
+        if (preferences.showNeighborInfo !== undefined) sets.push(`show_neighbor_info = ${bv(preferences.showNeighborInfo)}`);
+        if (preferences.showRoute !== undefined) sets.push(`show_route = ${bv(preferences.showRoute)}`);
+        if (preferences.showMotion !== undefined) sets.push(`show_motion = ${bv(preferences.showMotion)}`);
+        if (preferences.showMqttNodes !== undefined) sets.push(`show_mqtt_nodes = ${bv(preferences.showMqttNodes)}`);
+        if (preferences.showMeshCoreNodes !== undefined) sets.push(`show_meshcore_nodes = ${bv(preferences.showMeshCoreNodes)}`);
+        if (preferences.showAnimations !== undefined) sets.push(`show_animations = ${bv(preferences.showAnimations)}`);
+        if (preferences.showAccuracyRegions !== undefined) sets.push(`show_accuracy_regions = ${bv(preferences.showAccuracyRegions)}`);
+        if (preferences.showEstimatedPositions !== undefined) sets.push(`show_estimated_positions = ${bv(preferences.showEstimatedPositions)}`);
+        if (preferences.positionHistoryHours !== undefined) sets.push(`position_history_hours = ${preferences.positionHistoryHours ?? 'NULL'}`);
+
+        if (sets.length > 0) {
+          await this.executeRun(
+            sql.raw(`UPDATE user_map_preferences SET ${sets.join(', ')} WHERE ${userIdCol} = ${userId}`)
+          );
+        }
+      } else {
+        // INSERT — reference feature columns + createdAt/updatedAt (NOT NULL in all baselines)
+        // Quote createdAt/updatedAt for Postgres (case-sensitive identifiers)
+        // PG uses boolean type (true/false), SQLite/MySQL use integer (1/0)
+        const boolVal = (v: boolean | undefined, def: boolean) => {
+          const val = v !== undefined ? v : def;
+          return this.dbType === 'postgres' ? String(val) : (val ? '1' : '0');
+        };
+        const now = Date.now();
+        const q = this.dbType === 'postgres' ? '"' : '';
+        await this.executeRun(
+          sql.raw(`INSERT INTO user_map_preferences (
+            ${userIdCol}, map_tileset, show_paths, show_neighbor_info, show_route, show_motion,
+            show_mqtt_nodes, show_meshcore_nodes, show_animations, show_accuracy_regions,
+            show_estimated_positions, position_history_hours, ${q}createdAt${q}, ${q}updatedAt${q}
+          ) VALUES (
+            ${userId}, ${preferences.mapTileset ? `'${preferences.mapTileset.replace(/'/g, "''")}'` : 'NULL'},
+            ${boolVal(preferences.showPaths, false)}, ${boolVal(preferences.showNeighborInfo, false)},
+            ${boolVal(preferences.showRoute, true)}, ${boolVal(preferences.showMotion, true)},
+            ${boolVal(preferences.showMqttNodes, true)}, ${boolVal(preferences.showMeshCoreNodes, true)},
+            ${boolVal(preferences.showAnimations, false)}, ${boolVal(preferences.showAccuracyRegions, false)},
+            ${boolVal(preferences.showEstimatedPositions, true)}, ${preferences.positionHistoryHours ?? 'NULL'},
+            ${now}, ${now}
+          )`)
+        );
+      }
+    } catch (error) {
+      logger.error('[MiscRepository] Failed to save map preferences:', error);
+      throw error;
+    }
+  }
 }
 
 /**

--- a/src/db/schema/misc.ts
+++ b/src/db/schema/misc.ts
@@ -95,8 +95,8 @@ export const userMapPreferencesSqlite = sqliteTable('user_map_preferences', {
   centerLng: real('centerLng'),
   zoom: real('zoom'),
   selectedLayer: text('selectedLayer'),
-  createdAt: integer('createdAt').notNull(),
-  updatedAt: integer('updatedAt').notNull(),
+  createdAt: integer('createdAt'),
+  updatedAt: integer('updatedAt'),
 });
 
 export const userMapPreferencesPostgres = pgTable('user_map_preferences', {
@@ -106,8 +106,8 @@ export const userMapPreferencesPostgres = pgTable('user_map_preferences', {
   centerLng: pgReal('centerLng'),
   zoom: pgReal('zoom'),
   selectedLayer: pgText('selectedLayer'),
-  createdAt: pgBigint('createdAt', { mode: 'number' }).notNull(),
-  updatedAt: pgBigint('updatedAt', { mode: 'number' }).notNull(),
+  createdAt: pgBigint('createdAt', { mode: 'number' }),
+  updatedAt: pgBigint('updatedAt', { mode: 'number' }),
 });
 
 // ============ UPGRADE HISTORY ============
@@ -340,8 +340,8 @@ export const userMapPreferencesMysql = mysqlTable('user_map_preferences', {
   centerLng: myDouble('centerLng'),
   zoom: myDouble('zoom'),
   selectedLayer: myVarchar('selectedLayer', { length: 64 }),
-  createdAt: myBigint('createdAt', { mode: 'number' }).notNull(),
-  updatedAt: myBigint('updatedAt', { mode: 'number' }).notNull(),
+  createdAt: myBigint('createdAt', { mode: 'number' }),
+  updatedAt: myBigint('updatedAt', { mode: 'number' }),
 });
 
 export const upgradeHistoryMysql = mysqlTable('upgrade_history', {

--- a/src/server/migrations/007_add_missing_map_preference_columns.ts
+++ b/src/server/migrations/007_add_missing_map_preference_columns.ts
@@ -62,7 +62,11 @@ export async function runMigration083Postgres(client: PoolClient): Promise<void>
     { name: 'show_route', type: 'BOOLEAN DEFAULT true' },
     { name: 'show_motion', type: 'BOOLEAN DEFAULT true' },
     { name: 'show_mqtt_nodes', type: 'BOOLEAN DEFAULT true' },
+    { name: 'show_meshcore_nodes', type: 'BOOLEAN DEFAULT true' },
     { name: 'show_animations', type: 'BOOLEAN DEFAULT false' },
+    { name: 'show_accuracy_regions', type: 'BOOLEAN DEFAULT false' },
+    { name: 'show_estimated_positions', type: 'BOOLEAN DEFAULT false' },
+    { name: 'position_history_hours', type: 'INTEGER' },
     { name: 'created_at', type: 'BIGINT' },
   ];
 
@@ -92,7 +96,11 @@ export async function runMigration083Mysql(pool: MySQLPool): Promise<void> {
     { name: 'show_route', type: 'BOOLEAN DEFAULT true' },
     { name: 'show_motion', type: 'BOOLEAN DEFAULT true' },
     { name: 'show_mqtt_nodes', type: 'BOOLEAN DEFAULT true' },
+    { name: 'show_meshcore_nodes', type: 'BOOLEAN DEFAULT true' },
     { name: 'show_animations', type: 'BOOLEAN DEFAULT false' },
+    { name: 'show_accuracy_regions', type: 'BOOLEAN DEFAULT false' },
+    { name: 'show_estimated_positions', type: 'BOOLEAN DEFAULT false' },
+    { name: 'position_history_hours', type: 'INTEGER' },
     { name: 'created_at', type: 'BIGINT' },
   ];
 

--- a/src/server/models/User.ts
+++ b/src/server/models/User.ts
@@ -439,8 +439,6 @@ export class UserModel {
     showEstimatedPositions?: boolean;
     positionHistoryHours?: number | null;
   }): void {
-    const now = Date.now();
-
     // Check if preferences exist
     const existing = this.db.prepare('SELECT user_id FROM user_map_preferences WHERE user_id = ?').get(userId);
 
@@ -495,8 +493,6 @@ export class UserModel {
       }
 
       if (updates.length > 0) {
-        updates.push('updated_at = ?');
-        params.push(now);
         params.push(userId);
 
         const stmt = this.db.prepare(`
@@ -508,13 +504,15 @@ export class UserModel {
       }
     } else {
       // Insert new preferences
+      // Note: created_at/updated_at columns omitted — column naming is inconsistent
+      // across upgrade paths (createdAt vs created_at) and timestamps are not essential here
       const stmt = this.db.prepare(`
         INSERT INTO user_map_preferences (
           user_id, map_tileset, show_paths, show_neighbor_info,
           show_route, show_motion, show_mqtt_nodes, show_meshcore_nodes, show_animations,
           show_accuracy_regions, show_estimated_positions,
-          position_history_hours, created_at, updated_at
-        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
+          position_history_hours
+        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
       `);
 
       stmt.run(
@@ -529,9 +527,7 @@ export class UserModel {
         preferences.showAnimations ? 1 : 0,
         preferences.showAccuracyRegions ? 1 : 0, // default false
         preferences.showEstimatedPositions !== undefined ? (preferences.showEstimatedPositions ? 1 : 0) : 1, // default true
-        preferences.positionHistoryHours ?? null,
-        now,
-        now
+        preferences.positionHistoryHours ?? null
       );
     }
   }

--- a/src/services/database.ts
+++ b/src/services/database.ts
@@ -10005,55 +10005,14 @@ class DatabaseService {
   }
 
   /**
-   * Get user's map preferences - works with all database backends
+   * Get user's map preferences - delegates to MiscRepository (Drizzle ORM)
    */
   async getMapPreferencesAsync(userId: number): Promise<Record<string, any> | null> {
-    if (this.drizzleDbType === 'sqlite') {
-      return this.userModel.getMapPreferences(userId);
-    }
-
-    try {
-      const columns = `map_tileset, show_paths, show_neighbor_info, show_route, show_motion,
-        show_mqtt_nodes, show_meshcore_nodes, show_animations, show_accuracy_regions,
-        show_estimated_positions, position_history_hours`;
-
-      let row: any = null;
-
-      if (this.drizzleDbType === 'postgres' && this.postgresPool) {
-        const result = await this.postgresPool.query(
-          `SELECT ${columns} FROM user_map_preferences WHERE "userId" = $1`, [userId]
-        );
-        row = result.rows[0] || null;
-      } else if (this.drizzleDbType === 'mysql' && this.mysqlPool) {
-        const [rows] = await this.mysqlPool.query(
-          `SELECT ${columns} FROM user_map_preferences WHERE userId = ?`, [userId]
-        );
-        row = (rows as any[])[0] || null;
-      }
-
-      if (!row) return null;
-
-      return {
-        mapTileset: row.map_tileset || null,
-        showPaths: Boolean(row.show_paths),
-        showNeighborInfo: Boolean(row.show_neighbor_info),
-        showRoute: Boolean(row.show_route),
-        showMotion: Boolean(row.show_motion),
-        showMqttNodes: Boolean(row.show_mqtt_nodes),
-        showMeshCoreNodes: Boolean(row.show_meshcore_nodes),
-        showAnimations: Boolean(row.show_animations),
-        showAccuracyRegions: Boolean(row.show_accuracy_regions),
-        showEstimatedPositions: Boolean(row.show_estimated_positions),
-        positionHistoryHours: row.position_history_hours ?? null,
-      };
-    } catch (error) {
-      logger.error(`[DatabaseService] Failed to get map preferences async: ${error}`);
-      return null;
-    }
+    return this.miscRepo!.getMapPreferences(userId);
   }
 
   /**
-   * Save user's map preferences - works with all database backends
+   * Save user's map preferences - delegates to MiscRepository (Drizzle ORM)
    */
   async saveMapPreferencesAsync(userId: number, preferences: {
     mapTileset?: string;
@@ -10068,124 +10027,7 @@ class DatabaseService {
     showEstimatedPositions?: boolean;
     positionHistoryHours?: number | null;
   }): Promise<void> {
-    if (this.drizzleDbType === 'sqlite') {
-      this.userModel.saveMapPreferences(userId, preferences);
-      return;
-    }
-
-    const now = Date.now();
-
-    try {
-      // Check if row exists
-      let exists = false;
-      if (this.drizzleDbType === 'postgres' && this.postgresPool) {
-        const result = await this.postgresPool.query(
-          'SELECT "userId" FROM user_map_preferences WHERE "userId" = $1', [userId]
-        );
-        exists = (result.rows.length > 0);
-      } else if (this.drizzleDbType === 'mysql' && this.mysqlPool) {
-        const [rows] = await this.mysqlPool.query(
-          'SELECT userId FROM user_map_preferences WHERE userId = ?', [userId]
-        );
-        exists = ((rows as any[]).length > 0);
-      }
-
-      if (exists) {
-        // Build dynamic UPDATE
-        const updates: string[] = [];
-        const params: any[] = [];
-        let paramIdx = 1; // For Postgres $N placeholders
-
-        const fieldMap: Record<string, string> = {
-          mapTileset: 'map_tileset',
-          showPaths: 'show_paths',
-          showNeighborInfo: 'show_neighbor_info',
-          showRoute: 'show_route',
-          showMotion: 'show_motion',
-          showMqttNodes: 'show_mqtt_nodes',
-          showMeshCoreNodes: 'show_meshcore_nodes',
-          showAnimations: 'show_animations',
-          showAccuracyRegions: 'show_accuracy_regions',
-          showEstimatedPositions: 'show_estimated_positions',
-          positionHistoryHours: 'position_history_hours',
-        };
-
-        for (const [key, col] of Object.entries(fieldMap)) {
-          const value = (preferences as any)[key];
-          if (value !== undefined) {
-            if (this.drizzleDbType === 'postgres') {
-              updates.push(`${col} = $${paramIdx++}`);
-            } else {
-              updates.push(`${col} = ?`);
-            }
-            // Convert booleans for storage
-            if (typeof value === 'boolean') {
-              params.push(value);
-            } else {
-              params.push(value);
-            }
-          }
-        }
-
-        if (updates.length > 0) {
-          if (this.drizzleDbType === 'postgres') {
-            updates.push(`"updatedAt" = $${paramIdx++}`);
-            params.push(now);
-            const sql = `UPDATE user_map_preferences SET ${updates.join(', ')} WHERE "userId" = $${paramIdx}`;
-            params.push(userId);
-            await this.postgresPool!.query(sql, params);
-          } else if (this.drizzleDbType === 'mysql') {
-            updates.push('updatedAt = ?');
-            params.push(now);
-            const sql = `UPDATE user_map_preferences SET ${updates.join(', ')} WHERE userId = ?`;
-            params.push(userId);
-            await this.mysqlPool!.query(sql, params);
-          }
-        }
-      } else {
-        // INSERT new row
-        const boolVal = (v: boolean | undefined, def: boolean) => v !== undefined ? v : def;
-
-        if (this.drizzleDbType === 'postgres' && this.postgresPool) {
-          await this.postgresPool.query(
-            `INSERT INTO user_map_preferences (
-              "userId", map_tileset, show_paths, show_neighbor_info, show_route, show_motion,
-              show_mqtt_nodes, show_meshcore_nodes, show_animations, show_accuracy_regions,
-              show_estimated_positions, position_history_hours, created_at, "updatedAt"
-            ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14)`,
-            [
-              userId, preferences.mapTileset || null,
-              boolVal(preferences.showPaths, false), boolVal(preferences.showNeighborInfo, false),
-              boolVal(preferences.showRoute, true), boolVal(preferences.showMotion, true),
-              boolVal(preferences.showMqttNodes, true), boolVal(preferences.showMeshCoreNodes, true),
-              boolVal(preferences.showAnimations, true), boolVal(preferences.showAccuracyRegions, false),
-              boolVal(preferences.showEstimatedPositions, true), preferences.positionHistoryHours ?? null,
-              now, now,
-            ]
-          );
-        } else if (this.drizzleDbType === 'mysql' && this.mysqlPool) {
-          await this.mysqlPool.query(
-            `INSERT INTO user_map_preferences (
-              userId, map_tileset, show_paths, show_neighbor_info, show_route, show_motion,
-              show_mqtt_nodes, show_meshcore_nodes, show_animations, show_accuracy_regions,
-              show_estimated_positions, position_history_hours, created_at, updatedAt
-            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
-            [
-              userId, preferences.mapTileset || null,
-              boolVal(preferences.showPaths, false), boolVal(preferences.showNeighborInfo, false),
-              boolVal(preferences.showRoute, true), boolVal(preferences.showMotion, true),
-              boolVal(preferences.showMqttNodes, true), boolVal(preferences.showMeshCoreNodes, true),
-              boolVal(preferences.showAnimations, true), boolVal(preferences.showAccuracyRegions, false),
-              boolVal(preferences.showEstimatedPositions, true), preferences.positionHistoryHours ?? null,
-              now, now,
-            ]
-          );
-        }
-      }
-    } catch (error) {
-      logger.error(`[DatabaseService] Failed to save map preferences async: ${error}`);
-      throw error;
-    }
+    return this.miscRepo!.saveMapPreferences(userId, preferences);
   }
   // ============================================================
   // Async wrappers for sync methods (Phase 4 migration)

--- a/tests/api-exercise-test.sh
+++ b/tests/api-exercise-test.sh
@@ -511,6 +511,8 @@ echo ""
 echo -e "${BLUE}=== Map Preferences ===${NC}"
 
 check "GET /api/user/map-preferences" "$(api GET /api/user/map-preferences)" 200
+check "POST /api/user/map-preferences" "$(api POST /api/user/map-preferences '{"showRoute":true,"showNeighborInfo":false,"showMqttNodes":true}')" 200
+check "GET /api/user/map-preferences (after save)" "$(api GET /api/user/map-preferences)" 200
 
 # ─── Embed Profiles ───────────────────────────────────────
 

